### PR TITLE
OSDOCS#6222: Adding custom security groups in AWS

### PR DESCRIPTION
**Pull request by mjpytlak**
Version(s):
4.14+

Issue:
This issue addresses [osdocs-6222](https://issues.redhat.com/browse/OSDOCS-6222).

Link to docs preview:

- [AWS security groups](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-vpc.html#installation-aws-security-groups_installing-aws-vpc)
- Installing a cluster on AWS into an existing VPC > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-vpc.html#installation-aws-vpc-security-groups_installing-aws-vpc)
- Installing a private cluster on AWS > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-private.html#installation-aws-vpc-security-groups_installing-aws-private)
- Installing a cluster on AWS into a government region > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-government-region.html#installation-aws-vpc-security-groups_installing-aws-government-region)
- Installing a cluster on AWS into a Secret or Top Secret Region > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-secret-region.html#installation-aws-vpc-security-groups_installing-aws-secret-region)
- Installing a cluster on AWS China > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-china.html#installation-aws-vpc-security-groups_installing-aws-china-region)
- Installing a cluster using AWS Local Zone > [The edge compute pool for AWS Local Zones](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#machines-edge-machine-pool_installing-aws-localzone)
- Installing a cluster on AWS with remote workers on AWS Outposts > [Applying existing security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-outposts-remote-workers.html#installation-aws-vpc-security-groups_installing-aws-outposts-remote-workers)

QE review:
- [ ] QE has approved this change.

### CODE REVIEW

File: modules/installation-aws-vpc-security-groups.adoc [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#...
+compute:
+- hyperthreading: Enabled
+  name: worker
+  platform:
+    aws:
+      additionalSecurityGroupIDs:
+        - sg-1
+        - sg-2
+  replicas: 3
```
Comment (mtulio) [inline modules/installation-aws-vpc-security-groups.adoc:38]: I think it's important to make sure that feature also can work with the edge compute pool with Local Zones.
https://docs.openshift.com/container-platform/4.13/installing/installing_aws/installing-aws-localzone.html

@barbacbd can you see any restriction on that?

File: modules/installation-aws-vpc-security-groups.adoc [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  platform:
+    aws:
+      additionalSecurityGroupIDs:
+        - sg-3
+        - sg-4
+  replicas: 3
+platform:
+  aws:
+    region: us-east-1
+    subnets:
```
Comment (barbacbd) [inline modules/installation-aws-vpc-security-groups.adoc:51]: @mjpytlak did we want to note that the subnets are required or just let the user figure that out during install?

Comment (mjpytlak) [reply modules/installation-aws-vpc-security-groups.adoc]: > did we want to note that the subnets are required or just let the user figure that out during install?

@barbacbd I was originally thinking that given an earlier sample file in the topic states that the subnets are required, I would let it be. However - can't hurt. I updated code block with a few annotations. Outside of that, I made your proposed changes. Thanks for those. Please let me know when you have had a chance to look into the Local Zones case.

**Comment by mjpytlak** [comment]
@nirarg I am working with @barbacbd on the ability to apply existing AWS security groups when users BYON. I noticed that the AWS Outposts case requires users to BYON. The content I am updating is shared by the Outposts doc. Do either you or Brent have concerns about this content appear in that installation use case? [1]

[1] https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-outposts-remote-workers.html

**Comment by mjpytlak** [comment]
@barbacbd Following up on whether we can include the content on security groups to the Local Zones install use case. Do either you or @nirarg  have any concerns about adding this content to the AWS Outpost install use case?

File: modules/installation-aws-vpc-security-groups.adoc (hunk 1 above)
Comment (barbacbd) [reply modules/installation-aws-vpc-security-groups.adoc:37]: @mtulio as far as local zones go, security groups don't require anything special. The security groups just need to be attached to the VPC with the local zone subnet. Also we would require that these local zone subnets are provided in the subnets field of the install config.

Comment (barbacbd) [reply modules/installation-aws-vpc-security-groups.adoc:37]: @mtulio as far as edge computing, what concerns do you have? Security groups can be shared by many resources as they are just rules.

**Comment by barbacbd** [comment]
@mjpytlak @nirarg I don't see any issues with using custom security groups with any BYON. As long as the security groups are attached to a VPC then there should be no issues.

File: modules/installation-aws-vpc-security-groups.adoc (hunk 1 above)
Comment (mtulio) [reply modules/installation-aws-vpc-security-groups.adoc:37]: > @mtulio as far as edge computing, what concerns do you have? Security groups can be shared by many resources as they are just rules.

Yes, SGs can be assigned to any supported resource between the same VPC.

My concern was on the install side, I just wanna make sure the configuration below can be used - considering `edge` compute pool uses the same flow of `worker`, I expect only `sg-5` and `sg-6` in MachineSet manifests for Local Zones, just confirming it.

```yaml
compute:
- hyperthreading: Enabled
  name: worker
  platform:
    aws:
      additionalSecurityGroupIDs:
        - sg-1 <1>
        - sg-2
  replicas: 3
- hyperthreading: Enabled
  name: edge
  platform:
    aws:
      additionalSecurityGroupIDs:
        - sg-5
        - sg-6
```

Comment (barbacbd) [reply modules/installation-aws-vpc-security-groups.adoc:37]: Confirmed! The new machineset does show that it is for edge nodes and it has the security group that I mentioned in the install config and not the one used for the workers

**Comment by mjpytlak** [comment]
@mtulio I have added this content to the Local Zones doc. Notably I have updated the sample example to include local zone specific content. PTAL and let me know if we are OK as is or if updates are required. Thanks in advance.

- https://github.com/openshift/openshift-docs/pull/60921/files#diff-c1a8e33036504600b04232682d3adb385bd0cb256c8bb0fa17b50c6daec36635R38
- https://github.com/openshift/openshift-docs/pull/60921/files#diff-c1a8e33036504600b04232682d3adb385bd0cb256c8bb0fa17b50c6daec36635R65

File: modules/installation-applying-aws-security-groups.adoc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+[source,yaml]
+----
+#...
+compute:
+- hyperthreading: Enabled
+ifndef::local-zones[]
+  name: worker
+endif::local-zones[]
+ifdef::local-zones[]
+  - name: edge
```
[the remark this answers is not in the document]
Comment (mjpytlak) [reply modules/installation-applying-aws-security-groups.adoc:39]: I am interpreting your comment as `ifdef::local-zones[ ]` is incorrect. Are you referring to another hyphen in this example?

[the remark this answers is not in the document]
Comment (mtulio) [reply modules/installation-applying-aws-security-groups.adoc:39]: @mjpytlak - looking to the rendering page from this PR:
https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#installation-aws-vpc-security-groups_installing-aws-localzone

Screenshot from 2023-07-13 16-13-51

Comment (mtulio) [inline modules/installation-applying-aws-security-groups.adoc:39]:
```suggestion
compute:
- hyperthreading: Enabled
ifndef::local-zones[]
  name: worker
endif::local-zones[]
ifdef::local-zones[]
  name: edge
```

[the remark this answers is not in the document]
Comment (mtulio) [reply modules/installation-applying-aws-security-groups.adoc:39]: @mjpytlak Also looking at where that section is placed, I can see the docs is providing different examples of optional machine configuration (like Security Groups) here:
https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#machines-edge-machine-pool_installing-aws-localzone

- Configuration that uses an edge pool with default settings
- Configuration that uses an edge pool with a custom instance type
- Configuration that uses an edge pool with a custom EBS type

Do you think that optional security group deployment would fit in that list too, as it is also optional?

Comment (mjpytlak) [reply modules/installation-applying-aws-security-groups.adoc]: Done. Thank you for the clarification on Friday. I did not know the `-` from `hyperthreading` was inherited.

Comment (mtulio) [reply modules/installation-applying-aws-security-groups.adoc]: lgtm that part. Thanks

[the remark this answers is not in the document]
Comment (mtulio) [reply modules/installation-applying-aws-security-groups.adoc]: @mjpytlak ok, thanks,  I can see there. Do we need two examples about that topic on the Local Zone page? As you agreed to add the example[1] in the section, I can't see a benefit of [2] (as we already have detailed in the other pages). As a user [reading the rendered page] it seems the SG section ([2][Applying existing AWS security groups to the cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#installation-aws-vpc-security-groups_installing-aws-localzone)) is required in the installation process for Local Zones - which is not true.

- (1) Installing a cluster using AWS Local Zone > [The edge compute pool for AWS Local Zones](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#machines-edge-machine-pool_installing-aws-localzone)
- (2) Installing a cluster using AWS Local Zone > [Applying existing security groups to a cluster](https://60921--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#installation-aws-vpc-security-groups_installing-aws-localzone)

I am ok having only 1 on the Local Zones install page. Let me know your thoughts.

[the remark this answers is not in the document]
Comment (mjpytlak) [reply modules/installation-applying-aws-security-groups.adoc]: > I am ok having only 1 on the Local Zones install page. Let me know your thoughts.

@mtulio I agree and that was an oversight on my part. I removed Applying existing AWS security groups to the cluster.

[the remark this answers is not in the document]
Comment (mtulio) [reply modules/installation-applying-aws-security-groups.adoc]: @mjpytlak np at all! Thanks for revisiting.

**Comment by mjpytlak** [comment]
@yunjiang29 Draft docs ready for QE review. PTAL. Thank you.

**Comment by yunjiang29** [comment]
@gpei PTAL, thanks

File: modules/installation-applying-aws-security-groups.adoc [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+Applying existing AWS security groups to your control plane and compute machines can help you meet the security needs of your organization, in such cases where you need to control the incoming or outgoing traffic of these machines.
+
+.Prerequisites
+* You have created the security groups in AWS. For more information, see the AWS documentation about working with link:https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-security-groups.html[security groups].
+* The security groups must associated with the existing VPC to which you are deploying the cluster. The security groups cannot be associated with another VPC.
+* You have an existing `install-config.yaml` file.
+
+.Procedure
+
+. Edit the `compute.platform.aws.additionalSecurityGroupIDs` parameter to specify one or more custom security groups for your compute machines.
```
Comment (bergerhoffer) [inline modules/installation-applying-aws-security-groups.adoc:23]: Where are we editing, in the install-config.yaml file? I'd clarify that.

File: modules/installation-applying-aws-security-groups.adoc [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+. Edit `controlPlane.platform.aws.additionalSecurityGroupIDs` parameter to specify one or more custom security groups for your control plane machines.
```
Comment (bergerhoffer) [inline modules/installation-applying-aws-security-groups.adoc:24]:
```suggestion
. Edit the `controlPlane.platform.aws.additionalSecurityGroupIDs` parameter to specify one or more custom security groups for your control plane machines.
```

File: modules/installation-applying-aws-security-groups.adoc [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+. Save the file and reference it when deploying the cluster.
+
+.Sample `install-config.yaml` files that specifies custom security groups
```
Comment (bergerhoffer) [inline modules/installation-applying-aws-security-groups.adoc:27]:
```suggestion
.Sample `install-config.yaml` file that specifies custom security groups
```

File: modules/installation-applying-aws-security-groups.adoc [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+[source,yaml]
+----
+#...
```
Comment (bergerhoffer) [inline modules/installation-applying-aws-security-groups.adoc:30]: There should be a space, `# ...`

File: modules/installation-aws-security-groups.adoc [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+:localzone:
+endif::[]
+
+:_content-type: CONCEPT
+[id="installation-aws-security-groups_{context}"]
+= AWS security groups
+
+By default, the installation program creates and attaches security groups to control plane and compute machines. The rules associated with the default security groups cannot be modified.
+
+However, you can apply additional existing AWS security groups (custom security groups), which are associated with your existing VPC, to control plane and compute machines. Applying custom security groups can help you meet the security needs of your organization, in such cases where you need to control the incoming or outgoing traffic of these machines.
```
Comment (bergerhoffer) [inline modules/installation-aws-security-groups.adoc:21]: Not critical, but I'd argue that you don't need the `(custom security groups)`

**Review by bergerhoffer — COMMENTED** [review]
Nice job, just a few suggestions

**Comment by mjpytlak** [comment]
Appreciate the review @bergerhoffer. Always helpful.

**Pull request merged**